File: buildtools-rs/src/datadragon.rs
use std::{collections::HashMap, default};

use crate::fetch::FetchClient;
use json::JsonValue;
use serde::{Deserialize, Serialize};
use serde_json::Value;

type Version = String;

type Result<T> = std::result::Result<T, Box<dyn std::error::Error>>;
#[derive(Debug)]
pub enum DDragonError {
    Request(reqwest::Error),
    SerdeJson(serde_json::Error),
}

impl From<reqwest::Error> for DDragonError {
    fn from(value: reqwest::Error) -> Self {
        DDragonError::Request(value)
    }
}
impl From<serde_json::Error> for DDragonError {
    fn from(value: serde_json::Error) -> Self {
        DDragonError::SerdeJson(value)
    }
}

#[derive(Serialize, Deserialize, Debug, Default, Clone, PartialEq, Eq)]
#[serde(deny_unknown_fields)]
pub struct Realms {
    pub n: Named,
    pub v: String,            // "13.13.1"
    pub l: String,            // "en_US"
    pub cdn: String,          // "https://ddragon.leagueoflegends.com/cdn"
    pub dd: String,           // "13.13.1"
    pub lg: String,           // "13.13.1"
    pub css: String,          // "13.13.1"
    pub profileiconmax: i64,  // 28
    pub store: Option<Value>, // null
}

#[derive(Serialize, Deserialize, Debug, Default, Clone, Hash, PartialEq, Eq)]
#[serde(deny_unknown_fields)]
pub struct Named {
    pub item: String,        // "13.13.1"
    pub rune: String,        // "7.23.1"
    pub mastery: String,     // "7.23.1"
    pub summoner: String,    // "13.13.1"
    pub champion: String,    // "13.13.1"
    pub profileicon: String, // "13.13.1"
    pub map: String,         // "13.13.1"
    pub language: String,    // "13.13.1"
    pub sticker: String,     // "13.13.1"
}

pub fn get_realms() -> Result<Realms> {
    let region = "na";
    let url = format!("https://ddragon.leagueoflegends.com/realms/{region}.json");
    let body = reqwest::blocking::get(url)?.text()?;
    let realms: Realms = serde_json::from_str(&body)?;
    Ok(realms)
}

pub type ApiVersions = Vec<String>;
pub fn get_api_versions() -> Result<ApiVersions> {
    let url = "https://ddragon.leagueoflegends.com/api/versions.json";
    let body = reqwest::blocking::get(url)?.text()?;

    let v: ApiVersions = serde_json::from_str(&body)?;
    Ok(v)
}

pub fn get_latest_version() -> Result<String> {
    let realms = get_realms()?;
    Ok(realms.v)
}

pub struct DataDragon {
    client: FetchClient,
    version: String,
    lang: String,
    base_data_uri: String,
}
impl DataDragon {
    pub fn latest() -> Result<DataDragon> {
        let relms = get_realms()?;
        let version = relms.v;
        let lang = relms.l;
        let base_data_uri = format!("http://ddragon.leagueoflegends.com/cdn/{version}/data/{lang}");
        Ok(DataDragon {
            client: FetchClient::new()?,
            version,
            lang,
            base_data_uri,
        })
    }

    // The Raw Json File
    pub fn get_items_json(&mut self) -> Result<ItemJson> {
        let url = format!("{0}/item.json", self.base_data_uri);
        let body = self.client.fetch(url)?.text()?;

        // println!("--- START OF BODY ---");
        // println!("{body}");
        // println!("---- END OF BODY ----");

        let v: ItemJson = serde_json::from_str(&body)?;

        assert!(v.r#type == "item");
        assert!(v.version == self.version);
        Ok(v)
    }
    pub fn get_items(&mut self) -> Result<HashMap<String, Item>> {
        Ok(self.get_items_json()?.data)
    }
}

#[derive(Serialize, Deserialize, Debug, Default, Clone, PartialEq)]
// #[serde(deny_unknown_fields)]
pub struct ItemJson {
    pub r#type: String, // always 'item';
    pub version: Version,
    // pub basic: BasicItem,
    pub data: HashMap<String, Item>,
    // groups: HashMap<String, any>[],
    // tree: HashMap<String, any>[],
}

#[derive(Serialize, Deserialize, Debug, Default, Clone, PartialEq)]
#[serde(rename_all = "camelCase", default)]
pub struct BasicItem {
    pub name: String,

}
#[derive(Serialize, Deserialize, Debug, Default, Clone, PartialEq)]
#[serde(rename_all = "camelCase", default)]
#[serde(deny_unknown_fields)]
pub struct Item {
    pub name: String,
    pub description: String,
    #[serde(default = "default_colloq")]
    pub colloq: String,
    pub plaintext: String,
    pub special_recipe: i64,           // undefined defaults to 0
    pub consumed: bool,                // undefined defaults to false
    pub consume_on_full: Option<bool>, // undefined defaults to false
    #[serde(default = "default_1")]
    pub stacks: i64, // undefined defaults to 1
    #[serde(default = "default_true")]
    pub in_store: bool, // defaults to true
    pub hide_from_all: bool,           // defaults to false
    pub required_champion: String, //'FiddleSticks' | 'Kalista' | 'Sylas' | 'Gangplank', //defaults to ""
    pub required_ally: String,     // : 'Ornn', //defaults to ""
    pub from: Vec<String>,
    pub into: Vec<String>,
    pub image: Image,
    pub gold: Gold,
    pub tags: Vec<String>,
    pub maps: Value,
    // maps: { '11': boolean, '12': boolean, '21': boolean, '22': boolean },
    pub stats: HashMap<String, f64>,
    #[serde(default = "default_1")]
    depth: i64, // undefined defaults to 1
    pub effect: Value,
}
fn default_colloq() -> String {
    String::from(";")
}
fn default_1() -> i64 {
    1
}
fn default_true() -> bool {
    true
}

#[derive(Serialize, Deserialize, Debug, Default, Clone, PartialEq, Eq)]
#[serde(rename_all = "camelCase", default)]
#[serde(deny_unknown_fields)]
pub struct Gold {
    pub base: i64,
    pub purchasable: bool,
    pub total: i64,
    pub sell: i64,
}

#[derive(Serialize, Deserialize, Debug, Default, Clone, PartialEq, Eq)]
#[serde(rename_all = "camelCase", default)]
#[serde(deny_unknown_fields)]
pub struct Image {
   pub full: String,
   pub sprite: String,
   pub group: String,
   pub x: i64,
   pub y: i64,
   pub w: i64,
   pub h: i64,
  }

File: dev/datadragon.ts
import { Image } from '@/api/DataTypes';
import fetch from 'node-fetch';

export type Version = string;
export type Key = string;
export type Id = string;

export type Language =
  | 'cs_CZ' //Czech (Czech Republic)'
  | 'el_GR' //Greek (Greece)'
  | 'pl_PL' //Polish (Poland)'
  | 'ro_RO' //Romanian (Romania)'
  | 'hu_HU' //Hungarian (Hungary)'
  | 'en_GB' //English (United Kingdom)'
  | 'de_DE' //German (Germany)'
  | 'es_ES' //Spanish (Spain)'
  | 'it_IT' //Italian (Italy)'
  | 'fr_FR' //French (France)'
  | 'ja_JP' //Japanese (Japan)'
  | 'ko_KR' //Korean (Korea)'
  | 'es_MX' //Spanish (Mexico)'
  | 'es_AR' //Spanish (Argentina)'
  | 'pt_BR' //Portuguese (Brazil)'
  | 'en_US' //English (United States)'
  | 'en_AU' //English (Australia)'
  | 'ru_RU' //Russian (Russia)'
  | 'tr_TR' //Turkish (Turkey)'
  | 'ms_MY' //Malay (Malaysia)'
  | 'en_PH' //English (Republic of the Philippines)'
  | 'en_SG' //English (Singapore)'
  | 'th_TH' //Thai (Thailand)'
  | 'vn_VN' //Vietnamese (Viet Nam)'
  | 'id_ID' //Indonesian (Indonesia)'
  | 'zh_MY' //Chinese (Malaysia)'
  | 'zh_CN' //Chinese (China)'
  | 'zh_TW'; //Chinese (Taiwan)'
export class DataDragon {
  version: string;
  lang: Language;
  baseDataUri: string;

  /**
   * DataDragon class
   * @param {object} options options for DataDragon.
   * @param {string} [options.version="11.1.1"] version
   * @param {string} [options.lang="en_US"] language
   */

  constructor(options: { version?: Version; lang?: Language }) {
    this.version = options.version || '11.1.1';
    this.lang = options.lang || 'en_US';
    this.baseDataUri = `http://ddragon.leagueoflegends.com/cdn/${this.version}/data/${this.lang}`;
  }

  static async fromLatestVersion(): Promise<DataDragon> {
    const realms = await DataDragon.getRealms();
    return new DataDragon({ version: realms.v, lang: 'en_US' });
  }

  /**
   * Gets all availible versions
   * @returns {Promise<Version[]>} List of versions
   * @static
   */
  static async getVersions(): Promise<Version[]> {
    return fetch(`https://ddragon.leagueoflegends.com/api/versions.json`).then(
      (res) => res.json() as Promise<Version[]>,
    );
  }
  static async getLatestVersion(): Promise<Version> {
    return DataDragon.getVersions().then((x) => x[0]);
  }

  static async getRealms(): Promise<Realms> {
    return fetch(`https://ddragon.leagueoflegends.com/realms/na.json`).then(
      (res) => res.json() as Promise<Realms>,
    );
  }

  /**
   * Finds a champion
   * @param {string} champion champion name as a key AurelionSol.
   * @returns {Promise<ChampionComplex>}
   */

  // Ex. https://ddragon.leagueoflegends.com/cdn/10.12.1/data/en_US/champion/Aatrox.json
  async findChampion(champion: string): Promise<ChampionComplex> {
    const key = DataDragon.fixChampionKey(champion);
    return fetch(`${this.baseDataUri}/champion/${key}.json`)
      .then((res) => res.json() as Promise<NamedChampionJson>)
      .then((x) => x.data[key]);
  }

  static fixChampionKey(apiname: string): string {
    if (apiname === 'Wukong') return 'MonkeyKing';
    if (apiname === 'GnarBig') return 'Gnar';
    return apiname.replaceAll(/[' .]/g, '');
  }
  async getAllChampions(): Promise<ChampionBasicJson> {
    return fetch(`${this.baseDataUri}/champion.json`).then(
      (res) => res.json() as Promise<ChampionBasicJson>,
    );
  }
  async getAllChampionsFull(): Promise<ChampionFullJson> {
    return fetch(`${this.baseDataUri}/championFull.json`).then(
      (res) => res.json() as Promise<ChampionFullJson>,
    );
  }
  async getChallenges(): Promise<Record<string, any>[]> {
    return fetch(`${this.baseDataUri}/challenges.json`).then(
      (res) => res.json() as Promise<Record<string, any>[]>,
    );
  }
  // The Raw Json File
  async getItemsJson(): Promise<ItemJson> {
    return fetch(`${this.baseDataUri}/item.json`).then(
      (res) => res.json() as Promise<ItemJson>,
    );
  }
  // Mergeed with basic data to have all fields filled.
  async getItemsData(): Promise<Record<string, Item>> {
    const json = await this.getItemsJson();

    return json.data;
  }
}

export interface NamedChampionJson {
  type: 'champion';
  format: 'standAloneComplex';
  version: Version;
  data: Record<string, ChampionComplex>;
}

export interface ChampionFullJson {
  type: 'champion';
  format: 'full';
  version: Version;
  data: Record<string, ChampionComplex>;
  keys: Record<Key, Id>;
}

export interface ChampionBasicJson {
  type: 'champion';
  format: 'standAloneComplex';
  version: Version;
  data: Record<string, ChampionBasic>;
}

export interface ItemJson {
  type: 'item';
  version: Version;
  basic: Item;
  data: Record<string, Item>;
  groups: Record<string, any>[];
  tree: Record<string, any>[];
}
export interface Item {
  name: string;
  //rune: any;
  description: string;
  colloq: string;
  plaintext: string;
  specialRecipe?: number; // undefined defaults to 0
  consumed?: true; // undefined defaults to false
  consumeOnFull?: true; // undefined defaults to false
  stacks?: number; // undefined defaults to 1
  inStore?: false; // defaults to true
  hideFromAll?: true; // defaults to false
  requiredChampion?: 'FiddleSticks' | 'Kalista' | 'Sylas' | 'Gangplank'; //defaults to ""
  requiredAlly?: 'Ornn'; //defaults to ""
  from?: string[];
  into?: string[];
  image: Image;
  gold: { base: number; purchasable: boolean; total: number; sell: number };
  tags: string[];
  maps: { '11': boolean; '12': boolean; '21': boolean; '22': boolean };
  stats?: Record<string, number>;
  depth: number | undefined; // undefined defaults to 1
  effect?: any;
}

type RiotStatName =
  | 'FlatHPPoolMod'
  | 'rFlatHPModPerLevel'
  | 'FlatMPPoolMod'
  | 'rFlatMPModPerLevel'
  | 'PercentHPPoolMod'
  | 'PercentMPPoolMod'
  | 'FlatHPRegenMod'
  | 'rFlatHPRegenModPerLevel'
  | 'PercentHPRegenMod'
  | 'FlatMPRegenMod'
  | 'rFlatMPRegenModPerLevel'
  | 'PercentMPRegenMod'
  | 'FlatArmorMod'
  | 'rFlatArmorModPerLevel'
  | 'PercentArmorMod'
  | 'rFlatArmorPenetrationMod'
  | 'rFlatArmorPenetrationModPerLevel'
  | 'rPercentArmorPenetrationMod'
  | 'rPercentArmorPenetrationModPerLevel'
  | 'FlatPhysicalDamageMod'
  | 'rFlatPhysicalDamageModPerLevel'
  | 'PercentPhysicalDamageMod'
  | 'FlatMagicDamageMod'
  | 'rFlatMagicDamageModPerLevel'
  | 'PercentMagicDamageMod'
  | 'FlatMovementSpeedMod'
  | 'rFlatMovementSpeedModPerLevel'
  | 'PercentMovementSpeedMod'
  | 'rPercentMovementSpeedModPerLevel'
  | 'FlatAttackSpeedMod'
  | 'PercentAttackSpeedMod'
  | 'rPercentAttackSpeedModPerLevel'
  | 'rFlatDodgeMod'
  | 'rFlatDodgeModPerLevel'
  | 'PercentDodgeMod'
  | 'FlatCritChanceMod'
  | 'rFlatCritChanceModPerLevel'
  | 'PercentCritChanceMod'
  | 'FlatCritDamageMod'
  | 'rFlatCritDamageModPerLevel'
  | 'PercentCritDamageMod'
  | 'FlatBlockMod'
  | 'PercentBlockMod'
  | 'FlatSpellBlockMod'
  | 'rFlatSpellBlockModPerLevel'
  | 'PercentSpellBlockMod'
  | 'FlatEXPBonus'
  | 'PercentEXPBonus'
  | 'rPercentCooldownMod'
  | 'rPercentCooldownModPerLevel'
  | 'rFlatTimeDeadMod'
  | 'rFlatTimeDeadModPerLevel'
  | 'rPercentTimeDeadMod'
  | 'rPercentTimeDeadModPerLevel'
  | 'rFlatGoldPer10Mod'
  | 'rFlatMagicPenetrationMod'
  | 'rFlatMagicPenetrationModPerLevel'
  | 'rPercentMagicPenetrationMod'
  | 'rPercentMagicPenetrationModPerLevel'
  | 'FlatEnergyRegenMod'
  | 'rFlatEnergyRegenModPerLevel'
  | 'FlatEnergyPoolMod'
  | 'rFlatEnergyModPerLevel'
  | 'PercentLifeStealMod'
  | 'PercentSpellVampMod';

export interface ChampionBasic {
  version: Version;
  id: Id;
  key: Key;
  name: string;
  title: string;
  blurb: string;
  info: { attack: number; defense: number; magic: number; difficulty: number };
  image: Image;
  tags: string[];
  partype: string;
  stats: ChampionComplexStats;
}
export interface ChampionComplex {
  id: Id;
  key: Key;
  name: string;
  title: string;
  image: Image;
  skins: { id: string; num: number; name: string; chromas: boolean }[];
  lore: string;
  blurb: string;
  allytips: string[];
  enemytips: string[];
  tags: string[];
  partype: string;
  info: { attack: number; defense: number; magic: number; difficulty: number };
  stats: ChampionComplexStats;
  spells: ChampionComplexSpell[];
  passive: ChampionComplexPassive;
  recommended: [];
}

export interface ChampionComplexStats {
  hp: number;
  hpperlevel: number;
  mp: number;
  mpperlevel: number;
  movespeed: number;
  armor: number;
  armorperlevel: number;
  spellblock: number;
  spellblockperlevel: number;
  attackrange: number;
  hpregen: number;
  hpregenperlevel: number;
  mpregen: number;
  mpregenperlevel: number;
  crit: number;
  critperlevel: number;
  attackdamage: number;
  attackdamageperlevel: number;
  attackspeedperlevel: number;
  attackspeed: number;
}
export interface ChampionComplexPassive {
  name: string;
  description: string;
  image: Image;
}
export interface ChampionComplexSpell {
  id: string; // 'AatroxQ';
  name: string; // 'The Darkin Blade';
  description: string; // 'Aatrox slams his greatsword down, dealing physical damage. He can swing three times, each with a different area of effect.';
  tooltip: string; // 'Aatrox slams his greatsword, dealing <physicalDamage>{{ qdamage }} physical damage</physicalDamage>. If they are hit on the edge, they are briefly <status>Knocked Up</status> and they take <physicalDamage>{{ qedgedamage }}</physicalDamage> instead. This Ability can be <recast>Recast</recast> twice, each one changing shape and dealing 25% more damage than the previous one.{{ spellmodifierdescriptionappend }}';
  leveltip: {
    label: string[];
    effect: string[];
  };
  maxrank: number; // 5;
  cooldown: number[]; // [14, 12, 10, 8, 6];
  cooldownBurn: string; // '14/12/10/8/6';
  cost: number[]; // [0, 0, 0, 0, 0];
  costBurn: string; // '0';
  datavalues: {};
  effect: [
    null,
    [number, number, number, number, number],
    [number, number, number, number, number],
    [number, number, number, number, number],
    [number, number, number, number, number],
    [number, number, number, number, number],
    [number, number, number, number, number],
    [number, number, number, number, number],
    [number, number, number, number, number],
    [number, number, number, number, number],
    [number, number, number, number, number],
  ];
  effectBurn: [
    null,
    string,
    string,
    string,
    string,
    string,
    string,
    string,
    string,
    string,
    string,
  ];
  vars: never[];
  costType: string; // " {{ abilityresourcename }}",
  maxammo: string; //'-1';
  range: [number, number, number, number, number];
  rangeBurn: string; // '25000';
  image: Image;
  resource: string; // "{{ cost }} {{ abilityresourcename }}";
}

export interface Realms {
  n: {
    item: Version;
    rune: Version;
    mastery: Version;
    summoner: Version;
    champion: Version;
    profileicon: Version;
    map: Version;
    language: Version;
    sticker: Version;
  };
  v: Version;
  l: Language;
  cdn: 'https://ddragon.leagueoflegends.com/cdn';
  dd: Version;
  lg: Version;
  css: Version;
  profileiconmax: number;
  store: null;
}
